src/lib/proof-system/zkprogram.ts

```diff
@@ -161,12 +161,30 @@ const FeatureFlags = {
 };
 
 function createProgramState() {
+  let nonPureDataCache: Map<string, any[]> = new Map();
```

Comment:
we could just reuse the method cache for this too, would make it a bit cleaner but we would have to differenciate between nonpure auxiliary data and the auxiliary output its currently keeping track. both are valid imo